### ISSUE TITLE
White background (not black) when cropping images to larger dimensions

### ISSUE THREAD

**Issue by jenlampton** [backdrop/backdrop-issues#1946]
This has been an ongoing issue in Drupal since the 5.x cycle, see [this d.o issue](https://www.drupal.org/node/204497) for the backstory.

The current approach in 8.x is to add a configurable background color per image style. Though I agree this is a nice feature to have, I think we could solve the 80% use-case by making the color white (not black) by default. 

---

PR: https://github.com/backdrop/backdrop/pull/1698

**Comment by klonos** [comment]
...if I'm understanding the code correctly, this first tries to set the color to transparent and should that fail, *only then* it sets it to black(Drupal/current Backdrop) or white (PR 1428). Right?

Why is it that it fails in the first place to set it to transparent? Asking because that would solve the issue of having to set a specific color.

**Comment by jenlampton** [comment]
> Why is it that it fails in the first place to set it to transparent?

Not all image formats allow transparency. Jpg's, for example, would get a white background instead.

**Comment by klonos** [comment]
Oh, I thought that this was for the preview image used in the admin UI, in the image style edit form. That is always a .png AFAIK. That's why I asked. It makes sense for the actual image being uploaded for which we are to apply the style to. Ignore me :smile:

**Comment by Gormartsen** [comment]
@jenlampton I checked backdrop/backdrop#1428 test results.

There is error:
```
fail Background color was passed correctly image.test:372
```

image.test
```php
    $this->assertTrue(image_rotate_effect($this->image, array('degrees' => 90, 'bgcolor' => '#fff')), 'Function returned the expected value.');
    $this->assertToolkitOperationsCalled(array('rotate'));

    // Check the parameters.
    $calls = image_test_get_all_calls();
    $this->assertEqual($calls['rotate'][0][2], 0xffffff, 'Background color was passed correctly');
```
https://github.com/backdrop/backdrop/blob/230200112aff22cb2aee7cd9e1ce638a0cf60f64/core/modules/image/tests/image.test#L372

It's quite hard to find out what is `$calls['rotate'][0][2]` really mean here.

it is relate to:
image_test.module
https://github.com/backdrop/backdrop/blob/230200112aff22cb2aee7cd9e1ce638a0cf60f64/core/modules/simpletest/tests/image_test.module#L127
```php
/**
 * Image tookit's rotate operation.
 */
function image_test_rotate(stdClass $image, $degrees, $background = NULL) {
  _image_test_log_call('rotate', array($image, $degrees, $background));
  return TRUE;
}
```

So it is expecting and #ffffff background after image_rotate_effect function call

I assume mistake here:
https://github.com/backdrop/backdrop/pull/1428/files#diff-b2eebc799fc26976c4ff3d4293fbe853R520
```php
 // Assure consistency for the color string.
 +  $hex = '#' . $hex_color;
 +
 +  // Normalize shorthand versions.
 +  // '#FA3' will become '#FFAA33', '#FA37' will become '#FFAA3377'.
 +  if (strlen($hex_color) <= 4) {
 +    $hex_color = preg_replace('|([0-9A-F])|', '\1\1', $hex_color);
 +  }
```

You probably need to move ` $hex = '#' . $hex_color;` after checking if this one is 3 digit color:

```php
 +
 +  // Normalize shorthand versions.
 +  // '#FA3' will become '#FFAA33', '#FA37' will become '#FFAA3377'.
 +  if (strlen($hex_color) <= 4) {
 +    $hex_color = preg_replace('|([0-9A-F])|', '\1\1', $hex_color);
 +  }
 + // Assure consistency for the color string.
 +  $hex = '#' . $hex_color;
 
```

**Comment by jenlampton** [comment]
I don't think that has anything to do with failing tests, as they all pass locally for me. I've made the recommended change and pushed back up. We'll see what testbot says!

**Comment by jenlampton** [comment]
Still failing on Travis: https://travis-ci.org/jenlampton/backdrop/jobs/134942286

**Comment by Gormartsen** [comment]
Hi Jen!

Can you give me php info output from your localhost where tests pass?

My first proposed idea was not correct enough.

https://github.com/jenlampton/backdrop/commit/a9626b4d16b5f182f4550b4114820b314b534b25#diff-18d6dc07fe60b09d2236eb27fc771108L269

Here, after this like you need return to data[bgcolor] correct hex in 0x format

At least it was before here.





Regards,
Gor Martsen
@gormartsen

**Comment by jenlampton** [comment]
yep. 
![phpinfo](https://cloud.githubusercontent.com/assets/397895/15793044/3b26216c-2996-11e6-9cfa-1388042b582f.png)

**Comment by jenlampton** [comment]
You're right that the docs state that function should have a return value. I think that's less weird so I've made that change and pushed up the new code. (GD tests still passing locally) We'll see if it makes a difference to travis or gitci, but I'm doubtful.

**Comment by Gormartsen** [comment]
I saw tests drop the same error.

I will try to look into it.


Regards,
Gor Martsen
@gormartsen

**Comment by jenlampton** [comment]
Pushed a new PR that addresses @wesruv's feedback.

**Comment by quicksketch** [comment]
Posted a bunch of feedback to the PR: https://github.com/backdrop/backdrop/pull/1428

Overall the PR looks great. Nice new functionality that can be added in an unobtrusive way. Code needs some updates for consistency. Tests never did get passing, so maybe some more work is needed there also.

**Comment by jenlampton** [comment]
> Tests never did get passing, so maybe some more work is needed there also.

All tests still pass locally, I still think it's an infrastructure issue not a Backdrop-code issue. 

I rebased and pushed a new PR since I'm still needing this on most sites.

**Comment by quicksketch** [comment]
> All tests still pass locally, I still think it's an infrastructure issue not a Backdrop-code issue.

The same test fails locally for me as well. The nice thing is... that is much easier to debug. :)

I'm checking if I can get to the bottom of this.

**Comment by quicksketch** [comment]
The new PR is passing all tests. The problem apparently was rooted in loose type-checking. Apparently Jen's local version of PHP was checking `0xFFFFFF == '#FFFFFF'`, which was returning `TRUE` on that environment (but not mine or the testing infrastructure). I fixed the comparison to use matching input/output, and switched all the assert calls from `assertEqual()` to `assertIdentical()` so we get the same results on all environments.

I also gave it another review and found that `image_get_colorindex()` was in `image.inc`, but is GD-specific. As such, I renamed it to `_image_gd_get_colorindex()` and moved it into `image.gd.inc`.

Testing this out manually, everything works as expected. White is now used when cropping an image (unless it's a PNG, then it's transparent). Looks great and is ready to merge IMO.
